pkg/network/usm/utils/file_registry.go

```diff
@@ -40,170 +81,160 @@ func NewFileRegistry() *FileRegistry {
+		_ = deactivationCB(FilePath{ID: pathID})
+		// add `pathID` to blocklist so we don't attempt to re-register files
+		// that are problematic for some reason
 		r.blocklistByID[pathID] = struct{}{}
-		r.telemetry.libHookFailed.Add(1)
+		r.telemetry.fileHookFailed.Add(1)
 		return
 	}
 
-	reg := r.newRegistration(deactivate)
+	reg := r.newRegistration(deactivationCB)
 	r.byID[pathID] = reg
 	if len(r.byPID[pid]) == 0 {
 		r.byPID[pid] = pathIdentifierSet{}
 	}
 	r.byPID[pid][pathID] = struct{}{}
-	log.Debugf("registering library %s path %s by pid %d", pathID.String(), hostLibPath, pid)
-	r.telemetry.libRegistered.Add(1)
+	log.Debugf("registering file %s path %s by pid %d", pathID.String(), path.HostPath, pid)
+	r.telemetry.fileRegistered.Add(1)
 }
 
-// Unregister a pid if exists, unregisterCB will be called if his uniqueProcessesCount == 0
-func (r *FileRegistry) Unregister(pid int) {
-	pidU32 := uint32(pid)
-	r.m.RLock()
-	_, found := r.byPID[pidU32]
-	r.m.RUnlock()
+// Unregister a PID if it exists
+//
+// All files that were previously referenced by the given PID will have their
+// reference counters decremented by one. For any file for the number of
+// references drops to zero, we'll execute the *deactivationCB* previously
+// supplied during the `Register` call.
+func (r *FileRegistry) Unregister(pid uint32) {
+	r.m.Lock()
+	defer r.m.Unlock()
+	if r.stopped {
+		return
+	}
+
+	_, found := r.byPID[pid]
 	if !found {
 		return
 	}
 
-	r.m.Lock()
-	defer r.m.Unlock()
-	paths, found := r.byPID[pidU32]
+	paths, found := r.byPID[pid]
 	if !found {
 		return
 	}
 	for pathID := range paths {
 		reg, found := r.byID[pathID]
 		if !found {
-			r.telemetry.libUnregisterPathIDNotFound.Add(1)
+			r.telemetry.fileUnregisterPathIDNotFound.Add(1)
 			continue
 		}
 		if reg.unregisterPath(pathID) {
 			// we need to clean up our entries as there are no more processes using this ELF
 			delete(r.byID, pathID)
 		}
 	}
-	delete(r.byPID, pidU32)
+	delete(r.byPID, pid)
 }
 
-func (r *FileRegistry) GetRegisteredProcesses() map[int32]struct{} {
+// GetRegisteredProcesses returns a set with all PIDs currently being tracked by
+// the `FileRegistry`
+func (r *FileRegistry) GetRegisteredProcesses() map[uint32]struct{} {
 	r.m.RLock()
 	defer r.m.RUnlock()
 
-	result := make(map[int32]struct{}, len(r.byPID))
+	result := make(map[uint32]struct{}, len(r.byPID))
 	for pid := range r.byPID {
-		result[int32(pid)] = struct{}{}
+		result[pid] = struct{}{}
 	}
 	return result
 }
 
-// cleanup removes all registrations.
-// This function should be called in the termination, and after we're stopping all other goroutines.
-func (r *FileRegistry) Cleanup() {
-	for pathID, reg := range r.byID {
-		reg.unregisterPath(pathID)
+// Clear removes all registrations calling their deactivation callbacks
+// This function should be called once during in termination.
+func (r *FileRegistry) Clear() {
+	r.m.Lock()
+	defer r.m.Unlock()
+	if r.stopped {
+		return
 	}
```

Comment:
how that can happen?

Reply:
`Clear()` could be called multiple times